README.md
tct-shell
=========


debian/changelog
tct-shell (1.0.1) unstable; urgency=low

  * Packaged for Ubuntu

 -- Ed Bartosh <eduard.bartosh@intel.com>  Mon, 06 May 2013 14:37:15 +0300


pack.sh
#!/bin/bash
#
# Copyright (C) 2012 Intel Corporation
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
# Authors:
#             


# parse spec required name
NAME=`grep "Name:" tct-shell.spec | awk '{print $2}'`
if [ -z "$NAME" ];then
	echo "Name not specified in spec file"
	exit 1
fi

# parse spec required version
VERSION=`grep "Version:" tct-shell.spec | awk '{print $2}'`
if [ -z "$VERSION" ];then
	echo "Version not specified in spec file"
	exit 1
fi

SRC_ROOT=${PWD}
RPM_ROOT=/tmp/${NAME}_pack
ARCHIVE_TYPE=tar.gz	#tar.gz2
ARCHIVE_OPTION=czvf	#czvf

# check precondition
check_precondition()
{
	which $1 > /dev/null 2>&1
	if [ $? -ne 0 ]; then
		echo "Error: no tool: $1"
		exit 1
	fi
}
check_precondition rpmbuild
check_precondition make

# clean
echo "cleaning rpm workspace... >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>"
rm -rf $RPM_ROOT

# create workspace
echo "create rpm workspace... >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>"
mkdir -p $RPM_ROOT/RPMS $RPM_ROOT/SRPMS $RPM_ROOT/SPECS $RPM_ROOT/SOURCES $RPM_ROOT/BUILD $RPM_ROOT/src_tmp/$NAME-$VERSION

# prepare
echo "prepare workspace... >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>"
rm -rf *.rpm *.tar.bz2 *.tar.gz
cp -a $SRC_ROOT/* $RPM_ROOT/src_tmp/$NAME-$VERSION
# create Makefile in top src folder
#cp $SRC_ROOT/top_Makefile $RPM_ROOT/src_tmp/$NAME-$VERSION/Makefile
cp $PWD/tct-shell.spec $RPM_ROOT/SPECS
cd $RPM_ROOT/src_tmp
tar $ARCHIVE_OPTION $RPM_ROOT/SOURCES/$NAME-$VERSION.$ARCHIVE_TYPE $NAME-$VERSION
cd -

# build
echo "build from workspace... >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>"
cd  $RPM_ROOT/SPECS
rpmbuild -ba tct-shell.spec --clean --define "_topdir $RPM_ROOT" --target=noarch
cd -

# copy rpm
echo "copy from workspace... >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>"
echo "get rpm packages......"
find $RPM_ROOT -name "$NAME*.rpm" | grep -v debuginfo | xargs -n1 -I {} mv {} $PWD -f
echo "get $NAME-$VERSION.$ARCHIVE_TYPE......"
mv $RPM_ROOT/SOURCES/$NAME-$VERSION.$ARCHIVE_TYPE $PWD -f

# clean
echo "cleaning rpm workspace... >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>"
#rm -rf $RPM_ROOT

# validate
echo "checking result... >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>"
if [ -z "`ls | grep "\.rpm"`" ] || [ -z "`ls | grep "\.$ARCHIVE_TYPE"`" ];then
	echo "------------------------------ FAILED to build $NAME RPM packages --------------------------"
	exit 1
else
	echo "------------------------------ Done to build $NAME RPM packages --------------------------"
	ls *.rpm *.$ARCHIVE_TYPE 2>/dev/null
fi


style/application.js
function getScrollTop() {
	return f_scrollTop();
}

function f_scrollTop() {
	return f_filterResults($(window) ? $(window).scrollTop() : 0,
			document.documentElement ? document.documentElement.scrollTop : 0,
			document.body ? document.body.scrollTop : 0);
}
function f_filterResults(n_win, n_docel, n_body) {
	var n_result = n_win ? n_win : 0;
	if (n_docel && (!n_result || (n_result > n_docel)))
		n_result = n_docel;
	return n_body && (!n_result || (n_result > n_body)) ? n_body : n_result;
}

function setScrollTop() {
	$(window) ? $(window).scrollTop(0) : 0;
	document.documentElement ? document.documentElement.scrollTop = 0 : 0;
	document.body ? document.body.scrollTop = 0 : 0;
}

function goTopEx() {
	$node = $('#goTopBtn');
	if (getScrollTop() > 0) {
		$node.show();
	} else {
		$node.hide();
	}

	$(window).scroll(function() {
		if (getScrollTop() > 0) {
			$node.show();
		} else {
			$node.hide();
		}
	});

	$node.click(function() {
		setScrollTop();
	});
}

function drawRatio() {
	$('.suite_item').each(function(i, node) {
		drawSuiteRatio(node)
	});
}

$(".see_all").click(function(){
	$("#see_all").show();
    $("#see_fail").hide();
    $("#see_block").hide();
    $("#see_na").hide();
    updateToggles();
    return false;
});

$(".see_failed").click(function(){
	$("#see_all").hide();
    $("#see_fail").show();
    $("#see_block").hide();
    $("#see_na").hide();
    updateToggles();
    return false;
});

$(".see_blocked").click(function(){
	$("#see_all").hide();
    $("#see_fail").hide();
    $("#see_block").show();
    $("#see_na").hide();
    updateToggles();
    return false;
});

$(".see_na").click(function(){
	$("#see_all").hide();
    $("#see_fail").hide();
    $("#see_block").hide();
    $("#see_na").show();
    updateToggles();
    return false;
});

$("a.test_case_popup").click(function(){
	var $this = $(this);
	Popup.show($this.attr('id'));
    return false;
});

function drawSuiteRatio(node) {
	arrTitle = new Array("Passed", "Failed", "Blocked", "Not Executed");
	var $node = $(node);
	var $total = $node.find('.total');
	var $pass = $node.find('.pass');
	var $fail = $node.find('.fail');
	var $block = $node.find('.block');
	var $na = $node.find('.na');
	var $div = $node.find('.RatioGraphic');

	var total_int = parseInt($total.text());
	var pass_int = parseInt($pass.text());
	var fail_int = parseInt($fail.text());
	var block_int = parseInt($block.text());
	var na_int = parseInt($na.text());

	var pass_rate = pass_int * 100 / total_int;
	var fail_rate = fail_int * 100 / total_int;
	var block_rate = block_int * 100 / total_int;
	var na_rate = na_int * 100 / total_int;

	var areaWidth = 380;

	var pass_width = areaWidth * pass_rate / 100;
	var fail_width = areaWidth * fail_rate / 100;
	var block_width = areaWidth * block_rate / 100;
	var na_width = areaWidth * na_rate / 100;
	
	pass_rate = Math.round(pass_rate)
	fail_rate = Math.round(fail_rate)
	block_rate = Math.round(block_rate)
	na_rate = Math.round(na_rate)

	var pass_style = "padding:3px 0px 0px 0px;font-size:9pt;height:17px;text-align:center;color:white;font-weight:bold;background:url(&quot;blue.jpg&quot;);"
	var fail_style = "padding:3px 0px 0px 0px;font-size:9pt;height:17px;text-align:center;color:white;font-weight:bold;background:url(&quot;red.jpg&quot;);"
	var block_style = "padding:3px 0px 0px 0px;font-size:9pt;height:17px;text-align:center;color:white;font-weight:bold;background:url(&quot;orange.jpg&quot;);"
	var na_style = "padding:3px 0px 0px 0px;font-size:9pt;height:17px;text-align:center;color:white;font-weight:bold;background:url(&quot;gray.jpg&quot;);"

	var html = "<table width=\"380.68\" align=\"center\"><tbody><tr>";
	if (pass_width > 0){
	    html += "<td width=\""
			+ pass_width
			+ "\" style=\""
			+ pass_style
			+ "\" title=\"Passed :"
			+ pass_rate
			+ "%\">";
	    if (pass_width > 20){
	    	html += pass_rate + "%"
	    }
	    html +=  "</td>";
	}
	if (fail_width > 0){
		html += "<td width=\""
			+ fail_width
			+ "\" style=\""
			+ fail_style
			+ "\" title=\"Failed :"
			+ fail_rate
			+ "%\">";
	    if (fail_width > 20){
	    	html += fail_rate + "%"
	    }
	    html +=  "</td>";
	}
	if (block_width > 0){
		html += "<td width=\""
			+ block_width
			+ "\" style=\""
			+ block_style
			+ "\" title=\"Blocked :"
			+ block_rate
			+ "%\">";
	    if (block_width > 20){
	    	html += block_rate + "%"
	    }
	    html +=  "</td>";
	}
	if (na_width > 0){
		html += "<td width=\""
			+ na_width
			+ "\" style=\""
			+ na_style
			+ "\" title=\"Blocked :"
			+ na_rate
			+ "%\">";
	    if (na_width > 20){
	    	html += na_rate + "%"
	    }
	    html +=  "</td>";
	}
	html += "</tr></tbody></table>";
	$div.html(html);
}

tctshell/constants.py
#!/usr/bin/python
#
# Copyright (C) 2012 Intel Corporation
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
# Authors:
#              Tang, Shaofeng <shaofeng.tang@intel.com>

import ConfigParser
import os

CONFIG_FILE = "/opt/tct/shell/CONFIG"

parser = ConfigParser.ConfigParser()
parser.read(CONFIG_FILE)

class Constants:

     TEST_SUITE_DIR  = parser.get('TCTSHELL', 'TEST_SUITE_DIR')

     WRT_LAUNCHR_CMD = parser.get('TCTSHELL', 'WRT_LAUNCHR_CMD')

     RELEASED_SUITE_LOCATION = TEST_SUITE_DIR + "%s"

     TESTKIT_SHELL_HOME = parser.get('TCTSHELL', 'TCT_SHELL_HOME')

     TEMP_UNPASSED_XML_FOLDER = TESTKIT_SHELL_HOME + "/tmp/"

     TEMP_RESULT_XML_FOLDER = TESTKIT_SHELL_HOME + "/result/"

     if not TEMP_UNPASSED_XML_FOLDER.endswith('/'):
            TEMP_UNPASSED_XML_FOLDER = TEMP_UNPASSED_XML_FOLDER + "/"
     
     RERUN_OUTPUT = " -o " + TEMP_UNPASSED_XML_FOLDER + "result.xml"

     CAPABILITY_SUITE_NAME = "capability-tests"

     TCT_PLAN_FOLDER = parser.get('TCTSHELL', 'TCT_PLAN_FOLDER')

     STYLE_FOLDER = "/opt/tct/shell/style/*"

     CAPABILITY_PATH = "/opt/tct/shell/tmp/capability.xml"

     DEVICE_CAPABILITY_PATH = "/opt/usr/media/Documents/tct/capability.xml"

     #PARAM Option
     XMLFILE_PREFF = "-f device:"
     EXECUTE_PREFF = " -e "
     SKIP_NATIVE_MANUAL = "--non-active"
     ONLY_AUTO_CASES = "-A"
     ONLY_MANUAL_CASES = "-M"
     
     #SDB Command
     SDB_PUSH = "sdb push"
     SDB_SHELL = "sdb shell"
     SDB_DEVICES = "sdb devices"
     SDB_PULL = "sdb %s pull"

     
     #DEVICE
     DEVICE_TMP_FOLDER = parser.get('Device', 'DEVICE_TMP_FOLDER')
     DEVICE_TESTS_FILE = parser.get('Device', 'DEVICE_TESTS_FILE') + " "

     INSTALL_RPM       = "rpm -i --nodeps /tmp/%s"
     UNINSTALL_RPM     = "rpm -e --nodeps %s"
     REMOVE_RPM        = "rm -f /tmp/%s"

     #Plan Generator
     DEFAULT_MATCH_REGEX = "*.rpm"
     DEFAULT_UNMATCH_REGEX = None
     
     REGEX_FOR_SPLIT_PKG_NAME = "-\d\.\d\.\d"

     #Running Mode
     RUNNING_MODE_PLAN = "plan"
     RUNNING_MODE_RESULT = "result"
     RUNNING_MODE_SUITES = "suites"

     @staticmethod
     def add_right_enbrace(param):
         param = param.rstrip()
         param += "\""
         return param

     @staticmethod
     def indent(elem, level=0):
         i = "\n" + level*"  "
         if len(elem):
             if not elem.text or not elem.text.strip():
                 elem.text = i + "  "
             if not elem.tail or not elem.tail.strip():
                 elem.tail = i
             for elem in elem:
                 Constants.indent(elem, level+1)
             if not elem.tail or not elem.tail.strip():
                 elem.tail = i
         else:
             if level and (not elem.tail or not elem.tail.strip()):
                 elem.tail = i

     @staticmethod
     def copy_style_in_result_folder(result_folder):
         os.system("cp -r " + Constants.STYLE_FOLDER + " '" + result_folder + "'")
         print "Going to open the result report with FireFox, if Firefox is not available on your machine, please copy the folder '%s' to a Firefox machine." % result_folder
         os.system("firefox '%s'summary.xml &" % result_folder)




tctshell/plan_runner.py
#!/usr/bin/python
#
# Copyright (C) 2012 Intel Corporation
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
# Authors:
#              Tang, Shaofeng <shaofeng.tang@intel.com>

from constants import Constants
from xml.etree import ElementTree
import os
import glob
import re

class SuitePackage:
    def __init__(self, file_name, command):
        m = re.split(Constants.REGEX_FOR_SPLIT_PKG_NAME, file_name)
        self.name = m[0]
        self.command = command + " %s" % self.name

    def to_xml(self):
        suite = ElementTree.Element('suite')
        suite.set('name', self.name)
        command = ElementTree.Element('command')
        command.text = self.command
        suite.append(command)
        return suite

class PlanRunner:

    def __init__(self):
        self.suites = {}

    def load_local_repo(self, path, match, command, unmatch):
        try:
            os.chdir(path)
            print "match: %s" % match
            print "unmatch: %s" % unmatch
            for files in glob.glob(match):
                print "File name: %s" % str(files)
                if (unmatch is not None) and (re.search(unmatch, str(files))):
                    print "File %s is skipped with unmatch regex %s" % (str(files), unmatch)
                else:
                    suite = SuitePackage(files, command)
                    self.suites[suite.name] = suite
        except Exception, e:
            print "[ Error happen when reading the local repository, error: %s ]\n" % e

    def to_xml(self, xml_name):
        print "generating plan to %s" % xml_name
        root = ElementTree.Element('testplan')
        for suite_name in self.suites:
            suite = self.suites[suite_name]
            root.append(suite.to_xml())
        Constants.indent(root)
        tree = ElementTree.ElementTree()
        tree._setroot(root)
        tree.write(xml_name, encoding="utf-8")


tctshell/shellwrapper.py
#!/usr/bin/python
#
# Copyright (C) 2012 Intel Corporation
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
# Authors:
#              Tang, Shaofeng <shaofeng.tang@intel.com>


import os
import re
import sys, traceback
import time
import platform
import ctypes
import glob
import ConfigParser
import xml.etree.ElementTree as etree
from optparse import *
from shutil import copyfile
from tempfile import mktemp
from datetime import datetime
from constants import Constants
import os.path

#show all available suites
def show_available_suites(option, opt_str, value, parser):
    if platform.system() != "Linux":
        Constants.TEST_SUITE_DIR = "../suites"
    print "Test Suites:...\n"
    os.chdir(Constants.TEST_SUITE_DIR)
    rpms = sorted(glob.glob("*.rpm"))
    for files in rpms:
        #without ".rpm"
        print files[:-4]
    sys.exit(1)

# version option
def print_version(option, opt_str, value, parser):
    try:
        CONFIG = ConfigParser.ConfigParser()
        if platform.system() == "Linux":
            CONFIG.read('/opt/testkit/lite/VERSION')            
        else:
            VERSION_FILE = os.path.join(sys.path[0], 'VERSION')
            CONFIG.read(VERSION_FILE)
        VERSION = CONFIG.get('public_version', 'version')
        print "V%s" % VERSION
    except Exception, e:
        print "[ Error: fail to parse version info, error: %s ]\n" % e
    sys.exit(1)  

def print_planfolder(option, opt_str, value, parser):
    print "Plan folder: %s" % Constants.TCT_PLAN_FOLDER
    os.system("tree %s" % Constants.TCT_PLAN_FOLDER)
    sys.exit(1)

def print_resultfolder(option, opt_str, value, parser):
    print "Result folder: %s" % Constants.TEMP_RESULT_XML_FOLDER
    os.system("tree %s" % Constants.TEMP_RESULT_XML_FOLDER)
    sys.exit(1)

def invoke_sdb_devices(option, opt_str, value, parser):
    os.system(Constants.SDB_DEVICES)
    sys.exit(1)

def varnarg(option, opt_str, value, parser):
    """ parser srg"""
    value = []
    import re
    for arg in parser.rargs:
        if re.search('^--.+', arg) or re.search('^-[\D]', arg):
            break
        value.append(arg)

    del parser.rargs[:len(value)]
    setattr(parser.values, option.dest, value)

class TestKitShellWrapper:

     def __init__(self):
         print ""
         self._j = os.path.join
         self._e = os.path.exists
         self._d = os.path.dirname
         self._b = os.path.basename
         self._abspath = os.path.abspath
         self.testkit_dir = "/opt/testkit/lite"
         self.LOG_DIR = self.testkit_dir
         self.PID_FILE = self._j(self.LOG_DIR , "pid.log")
         if not platform.system() == "Linux":
             self.testkit_dir = self._d(self._abspath(__file__))
             sys.path += [self._j(self.testkit_dir)]
             self.testkit_dir = self._j(self.testkit_dir , "results")
         #self.clean_context()
         self.options = None
         self.options = None
         self.running_mode = None # plan suite result
         self.USAGE = "tct-shell [options] --testplan <somewhere/testplan.plan>\n\
examples: \n\
          tct-shell  --test package1 package2 ... packageN\n\
          tct-shell  --rerun-fail '<somewhere>/test-result.xml' \n\
          tct-shell  --testplan  <somewhere>/testplan.xml\n\
\n\
    run a test plan: \n\
          tct-shell  --testplan  <somewhere>/testplan.xml --output /tmp/wekit-tests-result.xml ...\n\
    run some test packages: \n\
          tct-shell  --test package1 package2 ... packageN --output /tmp/wekit-tests-result.xml ...\n\
    rerun all unpassed test: \n\
          tct-shell  --rerun-fail '<somewhere>/test-result.xml' ...\n\
    show all existed testplan which is in the folder (configured in /opt/tct/shell/CONFIG): \n\
          tct-shell  --plan-list\n\
    show all history result which is in the folder (configured in /opt/tct/shell/CONFIG): \n\
          tct-shell  --result-list\n\
    show all connected devices: \n\
          tct-shell  --device-list\n\
\n\
Note: \n\
          1) Proxy settings should be disabled when execute webapi packages\n\
          2) run command 'tct-shell', it might not be able to locate related module, run command 'export PYTHONPATH=/usr/lib/python2.7/site-packages' to resolve this issue"

     
     def print_usage(self):
         print self.USAGE

     def parse_options(self, argv):
         option_list = [
                    make_option("--testplan", dest="testplan_file",
                                action="callback", callback=varnarg,
                                help="Specify the testplan.xml."),
                    make_option("--output", dest="resultfile",
                                help="Specify output file for result xml. If more than one testxml provided, results will be merged together to this output file"),
                    make_option("--enable-memory-collection", dest="enable_memory_collection", action="store_true",
                                help="Enable the ability to release memory when the free memory is less than 100M"),
                    make_option("--version", dest="version_info", action="callback", callback=print_version, 
                                help="Show version information"),
                    make_option("--skip-iu", dest="skip_install", action="store_true", help="Automatically install and uninstall suite packages"),
                    make_option("--all-suites", dest="show_suites", action="callback", callback=show_available_suites, 
                                help="Show all available test-suites in the local repository, the local repository is defined in the configure '/opt/tct/shell/CONF'"),
                    #make_option("--capability", dest="capability_file", action="callback", callback=varnarg, 
                    #            help="Specify the capability file."),
                    make_option("--test", dest="suites", action="callback", callback=varnarg, 
                                help="Specify testing suites. If more than one suites are provided, just list them all and separate with whitespace"),
                    make_option("--rerun-fail", dest="fail_result_xml", action="callback", callback=varnarg, 
                                help="Rerun all fail testcase according to the specified result XML."),
                    make_option(Constants.SKIP_NATIVE_MANUAL, dest="skip_core_memory", action="store_true",
                                help="Disable the ability to set the result of core manual cases from the console"),
                    make_option("--deviceid", dest="deviceid", action="callback", callback=varnarg, 
                                help="set sdb device serial information."),
                    make_option("--plan-list", dest="show_plan_folder", action="callback", callback=print_planfolder, 
                                help="List all existed plan in the Plan folder. The plan folder is defined in the configure '/opt/tct/shell/CONF'"),
                    make_option("--result-list", dest="show_all_result", action="callback", callback=print_resultfolder, 
                                help="List all history results in the result folder. The result folder is defined in the configure '/opt/tct/shell/CONF'"),
                    make_option("--device-list", dest="show_all_device", action="callback", callback=invoke_sdb_devices, 
                                help="List all connected devices. just same with 'sdb devices'"),
                    make_option("--all", dest="all_tc", action="store_true", help="Both manual and auto test cases will be executed, without this option, only auto testcase will be executed."),
                    make_option("--manual", dest="only_manual", action="store_true", help="Only manual test cases will be executed"),
                    make_option("--id", dest="testcase_id", action="callback", callback=varnarg, 
                                help="Specify to run a test case by id."),
                  ]
         # detect non-params
         if len(argv) == 1:
             argv.append("--help")
         PARSERS = OptionParser(option_list=option_list, usage=self.USAGE)
         (self.options, args) = PARSERS.parse_args()
         if self.is_testplan_mode():

             conflicts = ["--testplan"]
             if self.options.fail_result_xml is not None:
                conflicts.append("--rerun-fail")
             if self.options.suites is not None:
                conflicts.append("--test") 
             self.conflict_exit(conflicts)
             
             self.running_mode = Constants.RUNNING_MODE_PLAN
         elif self.options.fail_result_xml is not None:
             
             conflicts = ["--rerun-fail"]
             if self.options.suites is not None:
                 conflicts.append("--test")
                 self.conflict_exit(conflicts)
             
             self.running_mode = Constants.RUNNING_MODE_RESULT
         elif self.options.suites is not None:
             self.running_mode = Constants.RUNNING_MODE_SUITES

         if self.options.all_tc and self.options.only_manual:
             conflicts = ["--all", "--manual"]
             self.conflict_exit(conflicts)

         self.check_args_number()

     def check_args_number(self):
         opt = ""
         if self.running_mode == Constants.RUNNING_MODE_PLAN and len(self.options.testplan_file) < 1:
             opt = "--testplan"
         elif self.running_mode == Constants.RUNNING_MODE_RESULT and len(self.options.fail_result_xml) < 1:
             opt = "--rerun-fail"
         elif self.running_mode == Constants.RUNNING_MODE_SUITES and len(self.options.suites) < 1:
             opt = "--test"
         #elif self.options.capability_file is not None and len(self.options.capability_file) < 1:
         #    opt = "--capability"
         elif self.options.deviceid is not None and len(self.options.deviceid) < 1:
             opt = "--deviceid"
         elif self.options.testcase_id is not None and len(self.options.testcase_id) < 1:
             opt = "--id"

         if len(opt) > 0:
             os.system("tct-shell -h")
             print "\ntct-shell: error: \"%s\" option requires an argument" % opt
             sys.exit(1)

             

     def conflict_exit(self, conflicts):
         if conflicts == None or len(conflicts) <= 1:
             return

         os.system("tct-shell -h")
         print "\ntct-shell: Conflicted options: %s" % conflicts
         sys.exit(1)

     def get_suite_param(self):
         param = Constants.XMLFILE_PREFF + "\""
         
         for suite_name in self.options.suites:
             param += "%s " % suite_name
         param = Constants.add_right_enbrace(param)

         param += Constants.EXECUTE_PREFF  + ("\" %s " % Constants.WRT_LAUNCHR_CMD)
         for e_name in self.options.suites:
             param += "%s " % e_name
         param = Constants.add_right_enbrace(param)
         return param

     def is_testplan_mode(self):
         return self.options.testplan_file is not None 

     def is_suite_mode(self):
         return self.options.suites is not None

     def is_fail_rerun_mode(self):
         return self.options.fail_result_xml is not None

     def get_plan_name(self):
         plan_name = ""
         if self.is_testplan_mode():
             plan_name = "Plan: %s" % self.options.testplan_file[0]
         elif self.is_suite_mode():
             plan_name = "Temporary plan to execute the suites: %s " % str(self.get_suites())
         elif self.is_fail_rerun_mode():
             plan_name = "Temporary plan to re-run all un-Passed testcases in the result_XML '%s' " %  self.options.fail_result_xml[0]
         return plan_name
  
     def is_auto_install(self):
         return not self.options.skip_install

     def get_output_param(self):
         return "-o %s" % self.get_output_file_name()

     def get_output_file_name(self):
         output = ""
         if self.options.resultfile is not None:
             output = os.path.abspath(self.options.resultfile)
         else:
             output = Constants.TEMP_UNPASSED_XML_FOLDER + "result.xml"
         return output

     def get_memory_param(self):
         memory_param = ""
         if self.options.enable_memory_collection:
             memory_param = "--enable-memory-collection"
         return memory_param

     def get_skip_native_manual_cases_param(self):
         skip = ""
         if self.options.skip_core_memory:
             skip = Constants.SKIP_NATIVE_MANUAL
         return skip

     def get_deviceid_param(self):
         deviceid = ""
         if self.options.deviceid is not None:
             deviceid = " --deviceid %s" % self.options.deviceid[0]
         return deviceid

     def get_sdb_device_id_param(self):
         deviceid = ""
         if self.options.deviceid is not None:
             deviceid = " -s %s" % self.options.deviceid[0]
         return deviceid

     def get_capability_param(self):
         capability_file = ""
         if os.path.isfile(Constants.CAPABILITY_PATH):
             capability_file = " --capability %s" % Constants.CAPABILITY_PATH
         return capability_file 
     
     def is_cap_param_available(self):
         return os.path.isfile(Constants.CAPABILITY_PATH)

     def get_auto_case_param(self):
         auto = Constants.ONLY_AUTO_CASES
         if self.options.all_tc or self.options.only_manual:
             auto = ""
         return auto

     def get_manual_case_param(self):
         manual = ""
         if self.options.only_manual:
             manual = Constants.ONLY_MANUAL_CASES
         return manual

     def get_testcase_id(self):
         tc_id = ""
         if self.options.testcase_id is not None:
             tc_id = " --id %s" % self.options.testcase_id[0]
         return tc_id
     
     def print_result_summary(self, result_file):         
         try:
             tree = etree.parse(result_file)
             root = tree.getroot()
             suites = root.findall('suite')
             
             total_cases = len(root.findall('suite/set/testcase'))
             print "[Result: execute %d suites]" % len(suites)
             print "[  total case number: %d ]" % total_cases
             if total_cases > 0:
                 print "[    Pass Rate: %d" % int(len(root.findall('suite/set/testcase[@result="PASS"]')) * 100 /total_cases) + "% ]"
             else:
                 print "[    Pass Rate: 0% ]"
             print "[    PASS case number: %d ]" % len(root.findall('suite/set/testcase[@result="PASS"]'))
             print "[    FAIL case number: %d ]" % len(root.findall('suite/set/testcase[@result="FAIL"]'))
             print "[    BLOCK case number: %d ]" % len(root.findall('suite/set/testcase[@result="BLOCK"]'))
             print "[    N/A case number: %d ]" % len(root.findall('suite/set/testcase[@result="N/A"]'))
             for suite in suites:
                 print " Suite: %s" % suite.get("name")
                 suite_cases_number = len(suite.findall('set/testcase'))
                 print "      |---total case number: %d " % suite_cases_number
                 if suite_cases_number > 0 :
                     print "      |      |---Pass rate: %d" % int(len(suite.findall('set/testcase[@result="PASS"]')) * 100 /len(suite.findall('set/testcase'))) + "%"
                 else:
                     print "      |      |---Pass rate: 0%"
                 print "      |      |---PASS case number: %d " % len(suite.findall('set/testcase[@result="PASS"]'))
                 print "      |      |---FAIL case number: %d " % len(suite.findall('set/testcase[@result="FAIL"]'))
                 print "      |      |---BLOCK case number: %d " % len(suite.findall('set/testcase[@result="BLOCK"]'))
                 print "      |      |---N/A case number: %d " % len(suite.findall('set/testcase[@result="N/A"]'))
         except Exception, e:
            print "[ Error: When calculating the summary of Result XML, error: %s ]\n" % e

     def get_suites(self):
         suites = []
         if self.options.suites:
             for suite_str in self.options.suites:
                 suite_str_array = suite_str.split(" ")
                 for suite_name in suite_str_array:
                     suites.append(suite_name)
         return suites
     
     def check_sdb_devices(self):
         sdb_ok = False
         import subprocess

         proc = subprocess.Popen(["sdb devices"], stdout=subprocess.PIPE, shell=True)
         (out, err) = proc.communicate()
         print "sdb devices: %s" % out
         
         if self.options.deviceid is not None and self.options.deviceid[0] is not None:
             if self.options.deviceid[0] in out:
                 sdb_ok = True
                 print "The specified device %s is connected. Going to start testing" % self.options.deviceid[0]
             else:
                 print "The specified device %s is not connected. exiting" % self.options.deviceid[0]
         elif "device-1" in out:
             sdb_ok = True
             print "No device is specified, use the only connected USB device."
         else:
             print "No device is connected. exiting"

         return sdb_ok

